Opening a scheduled transaction to view causes app to crash

The scheduled transaction list is displayed, but the individual items cannot be opened for editing.. 
This behaviour currently occurs on `develop` branch.

This can be fixed and extended so that tapping on a scheduled transaction does not directly go to editing it, but rather first displays details of it using `TransactionDetailFragment`. And then from there, it can be edited further.

The crash happens when the scheduled transaction doesn't have an end date.

When we don't set an end date, no value is stored in the corresponding field in the database. This is then read as a null value in `RecurrenceDbAdapter.buildModelInstance` into `ScheduledAction.mRecurrence.mPeriodEnd`, which provokes a NullPointerException in `Recurrence.getCount()`.

08-03 18:03:52.791 17435-17435/org.gnucash.android.devel E/AndroidRuntime: FATAL EXCEPTION: main
   Process: org.gnucash.android.devel, PID: 17435
   java.lang.RuntimeException: Unable to start activity ComponentInfo{org.gnucash.android.devel/org.gnucash.android.ui.common.FormActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'long java.sql.Timestamp.getTime()' on a null object reference
       at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2455)
       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2519)
       at android.app.ActivityThread.access$800(ActivityThread.java:162)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1412)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loop(Looper.java:189)
       at android.app.ActivityThread.main(ActivityThread.java:5532)
       at java.lang.reflect.Method.invoke(Native Method)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:950)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:745)
    Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'long java.sql.Timestamp.getTime()' on a null object reference
       at org.gnucash.android.model.Recurrence.getCount(Recurrence.java:268)
       at org.gnucash.android.model.Recurrence.getRuleString(Recurrence.java:164)
       at org.gnucash.android.model.ScheduledAction.getRuleString(ScheduledAction.java:407)
       at org.gnucash.android.ui.transaction.TransactionFormFragment.initializeViewsWithTransaction(TransactionFormFragment.java:503)
       at org.gnucash.android.ui.transaction.TransactionFormFragment.onActivityCreated(TransactionFormFragment.java:355)
       at android.support.v4.app.Fragment.performActivityCreated(Fragment.java:1983)
       at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1092)
       at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1252)
       at android.support.v4.app.BackStackRecord.run(BackStackRecord.java:742)
       at android.support.v4.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:1617)
       at android.support.v4.app.FragmentController.execPendingActions(FragmentController.java:339)
       at android.support.v4.app.FragmentActivity.onStart(FragmentActivity.java:602)
       at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1223)
       at android.app.Activity.performStart(Activity.java:5982)
       at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2418)
       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2519) 
       at android.app.ActivityThread.access$800(ActivityThread.java:162) 
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1412) 
       at android.os.Handler.dispatchMessage(Handler.java:106) 
       at android.os.Looper.loop(Looper.java:189) 
       at android.app.ActivityThread.main(ActivityThread.java:5532) 
       at java.lang.reflect.Method.invoke(Native Method) 
       at java.lang.reflect.Method.invoke(Method.java:372) 
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:950) 
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:745) 

--- a/app/src/main/java/org/gnucash/android/model/Recurrence.java
+++ b/app/src/main/java/org/gnucash/android/model/Recurrence.java
@@ -263,6 +263,9 @@ public class Recurrence extends BaseModel {
      * @return Number of occurrences, or -1 if there is no end date
      */
     public int getCount(){
+        if (mPeriodEnd == null)
+            return -1;
+
         int count = 0;
         LocalDate startDate = new LocalDate(mPeriodStart.getTime());
         LocalDate endDate = new LocalDate(mPeriodEnd.getTime());

--- a/app/src/test/java/org/gnucash/android/test/unit/model/RecurrenceTest.java
+++ b/app/src/test/java/org/gnucash/android/test/unit/model/RecurrenceTest.java
@@ -58,4 +58,19 @@ public class RecurrenceTest {
 
         assertThat(recurrence.getCount()).isEqualTo(10);
     }
+
+    /**
+     * When no end period is set, getCount() should return the special value -1.
+     *
+     * <p>Tests for bug https://github.com/codinguser/gnucash-android/issues/526</p>
+     */
+    @Test
+    public void notSettingEndDate_shouldReturnSpecialCountValue() {
+        Recurrence recurrence = new Recurrence(PeriodType.MONTH);
+        Calendar cal = Calendar.getInstance();
+        cal.set(2015, Calendar.OCTOBER, 5);
+        recurrence.setPeriodStart(new Timestamp(cal.getTimeInMillis()));
+
+        assertThat(recurrence.getCount()).isEqualTo(-1);
+    }
 }

